File: arrays-Hashing/leetCode_49.py
"""
Given an array of strings strs, group the 
anagrams
 together. You can return the answer in any order.

 

Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]

Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
"""

class Solution:
    def groupAnagrams(self,strings):
        hashMap = {}
        for string in strings:
            key = "".join(sorted(string))
            if key in hashMap:
                hashMap[key].append(string)
            else:
                hashMap[key] = [string]
        return hashMap.values()

s = Solution()
print(s.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

File: arrays-Hashing/leetCode_49.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.List;

public class leetCode_49 {
    public List<List<String>> groupAnagrams (String[] strs){
        Map<String, List<String>> wordMap = new HashMap<>();
        for (int i = 0; i < strs.length; i++){
            char[] array = strs[i].toCharArray();
            Arrays.sort(array);
            String str = new String(array);
            if (!wordMap.containsKey(str)){
                wordMap.put(str, new ArrayList<>());
            }
            wordMap.get(str).add(strs[i]);
        }
        return new ArrayList<>(wordMap.values());
    }

    public static void main(String[] args) {
        leetCode_49 solution = new leetCode_49();
        String[] strings = {"eat","tea","tan","ate","nat","bat"};
        List<List<String>> result = solution.groupAnagrams(strings);
        System.out.println("The result is :" + result);
    }
}
